File: db-objects/src/main/java/com/simplesys/container/OrdDoc.java
package com.simplesys.container;

import oracle.jdbc.OracleData;
import oracle.jdbc.OracleDataFactory;
import oracle.sql.CHAR;
import oracle.sql.CLOB;
import oracle.sql.NUMBER;
import oracle.sql.STRUCT;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.Reader;
import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Struct;

public class OrdDoc implements OracleData, OracleDataFactory {

  public OrdSource source;
  public CHAR format;
  public CHAR mimeType;
  public NUMBER contentLength;
  public CLOB comments;

  static final OrdDoc _ordDoc = new OrdDoc();

  public static OracleDataFactory getOracleDataFactory() {
    return _ordDoc;
  }

  public OrdDoc() {
  }

  public OrdDoc(OrdSource source,
                CHAR format,
                CHAR mimeType,
                NUMBER contentLength,
                CLOB comments) {
    this.source = source;
    this.format = format;
    this.mimeType = mimeType;
    this.contentLength = contentLength;
    this.comments = comments;
  }


  @Override
  public Object toJDBCObject(Connection connection) throws SQLException {
    Object[] attributes = {source, format, mimeType, contentLength, comments};
    Struct struct = connection.createStruct("ORDSYS.ORDDOC", attributes);
    return struct;
  }

  @Override
  public OracleData create(Object o, int i) throws SQLException {
    if (o == null) return null;

    Object[] attributes = ((STRUCT) o).getOracleAttributes();
    return new OrdDoc(
            //(OrdSource) attributes[0],
            null,
            (CHAR) attributes[1],
            (CHAR) attributes[2],
            (NUMBER) attributes[3],
            (CLOB) attributes[4]
    );
  }
}

File: db-objects/src/main/scala/com/simplesys/container/scala/OrdDoc.scala
package com.simplesys.container.scala

import com.simplesys.container.Helper
import com.simplesys.container.java.JOrdDoc
import oracle.jdbc.{OracleClob, OracleConnection}
import oracle.sql.{CLOB, NUMBER}

object OrdDoc {
    implicit def JOrdSource2SOrdSource(ordDoc: JOrdDoc): OrdDoc = new OrdDoc {
        override val source: Option[OrdSource] = Option(ordDoc.source).map(OrdSource.JOrdSource2SOrdSource(_))
        override val format: Option[String] = Option(ordDoc.format).map(Helper.asString(_))
        override val mimeType: Option[String] = Option(ordDoc.mimeType).map(Helper.asString(_))
        override val contentLength: Option[BigDecimal] = Option(ordDoc.contentLength).map(Helper.asBigDecimal(_))
        override val comments: Option[String] = Option(Helper.clobToString(ordDoc.comments))
    }

    implicit def SOrdSource2JOrdSource(ordDoc: OrdDoc)(implicit connection: OracleConnection): JOrdDoc = {
        val res = new JOrdDoc
        ordDoc.source.foreach (res.source = _)
        ordDoc.format.foreach(format ⇒ res.format = Helper.asCHAR(format))
        ordDoc.mimeType.foreach(mimeType ⇒ res.mimeType = Helper.asCHAR(mimeType))
        ordDoc.contentLength.foreach(contentLength ⇒ res.contentLength = new NUMBER(contentLength.bigDecimal))
        ordDoc.comments.foreach(comment ⇒ res.comments = {
            val clob = connection.createClob().asInstanceOf[OracleClob]
            clob.setString(1, comment)
            clob
        })
        res
    }
}

trait OrdDoc {
    val source: Option[OrdSource]
    val format: Option[String]
    val mimeType: Option[String]
    val contentLength: Option[BigDecimal]
    val comments: Option[String]
}
